### DISCUSSION TITLE
Best practice for returning undefined data

### DISCUSSION THREAD

**Discussion by robingenz** [TanStack/query#7020] [Q&A] [answered]
First of all, thank you for this great project!

I currently get the following error message with Angular:

```
Query data cannot be undefined. Please make sure to return a value other than undefined from your query function. Affected query key: ["tasks","ldPVmyj3qZaVrfmGVOsmG"]
```

This is my code:

```ts
public getTaskById(
    id: string,
  ): Result<QueryObserverResult<Task | undefined, Error>> {
    return this.#query({
      queryKey: ['tasks', id],
      queryFn: () => this.tasksService.getTaskById(id),
    });
  }
```

As you can see, I am searching an entity `tasks` for a task with a specific id. If this item does not exist, `this.tasksService.getTaskById(...)` returns `undefined`. However, this leads to the above error message. 

I understand the error message. However, I wonder what the best practices are to implement this.

**Comment by riccardoperra** [comment]
Not definitely an expert but I'm also curious to know the answer. Anyway in my opinion:
- return `null` instead of undefined, which seems not allowed as a value for the cache key.
- api should return an error if the data not exists. If it doesn't, it should be handled manually
```ts
{ 
  queryKey: ['tasks', id]
  queryFn: () => this.taskService.getTaskById(id).then((data) => {
     if (data) { 
       throw new MyCustomError()
     }
     return data
  }),
  retry? // ... handle logic for MyCustomError
  throwOnError? // handle logic for MyCustomError
}
````

**Comment by TkDodo** [comment, marked as answer]
if you want the query go to into error state when the item does not exist, you can check and throw a custom error as the answer by @riccardoperra shows.

if you're fine with having no data for a successfully resolved query, I'd just transform to `null` because that is allowed:

```
public getTaskById(
    id: string,
  ): Result<QueryObserverResult<Task | null, Error>> {
    return this.#query({
      queryKey: ['tasks', id],
      queryFn: async () => {
        const task = this.tasksService.getTaskById(id)
        return task ?? null
      }
    });
  }
```

of course you could also change `getTaskById` to already return `null`.
